### ISSUE TITLE
Low-zoom coastline simplification and outline problems

### ISSUE THREAD

**Issue by jeisenbe** [gravitystorm/openstreetmap-carto#3695]
### Current Situation

Since PR #3065 the land polygons have been simplified and outlined at z0 to z7. Previously, the land polygons were shown without an outline and without any simplification.

The simplified land polygon shapefiles are used, which have nearly full detail for all land features at z9 and below, but Mapnik's line-simplify option is used to somewhat smooth out the coastline, and then a sub-pixel-width outline is added.

The 0.5 pixel outline is currently offset 0.5 pixels from the coast, which leads to some errors when exporting at higher resolution, when the coastline is particularly curvy. There is also a mismatch between small islands and the outline; some small islands are shown but do not have a visible outline. 

When exported at 2x or higher resolution, these problems are even more apparent.

### Options for resolving the issue

1) Revert #3065 - remove simplification of the water polygons and the outlines

2) Switch to generalized coastline shapefiles for the outline and generalized water polygons for the z0 to z7 land / oceans
- This would allow rendering the coastline outline, but will have more consistent results and will also work when exporting at higher resolutions

### Links and screenshots illustrating the problem

Current rendering of Greece, screenshot:
https://www.openstreetmap.org/#map=5/38.065/24.038
![z5-greece-screenshot](https://user-images.githubusercontent.com/42757252/53300146-b4f82080-3875-11e9-80f1-e0908ed4cec7.png)

Land polygons only, current master branch:
![z5-greece-master-missing-islands](https://user-images.githubusercontent.com/42757252/53300116-5e8ae200-3875-11e9-84d4-eebcbbd0057d.png)

Zoomed in; missing island outlines are apparent
![z5-greece-missing-islands-zoomin](https://user-images.githubusercontent.com/42757252/53300118-63e82c80-3875-11e9-8bef-e265ff983654.png)

Exported at 2x from kosmtik:
![master-export2x-greece](https://user-images.githubusercontent.com/42757252/53300162-01436080-3876-11e9-8ac3-5c4b95c596c6.png)

Exported at 5x from kosmtik:
![master-export5x-greece](https://user-images.githubusercontent.com/42757252/53300164-086a6e80-3876-11e9-9e69-91f0a2224ff1.png)

(Compared to rendering with generalized coastline shapefiles)
![z5-export2x-greece-generalized-coastlines-1zlate](https://user-images.githubusercontent.com/42757252/53300207-9e9e9480-3876-11e9-9e41-3e714cdc85e5.png)

**Comment by jeisenbe** [comment]
Is everyone happy with the simple solution of removing the outline and simplification in PR #3694 - or should I show some more examples of rendering with the generalized coastlines at z0 to z7?

**Comment by jeisenbe** [comment]
The outline also over-emphasizes tidal channels in mangroves and saltwater marshes, and river mouths, eg:

z7 Timika current rendering
https://www.openstreetmap.org/#map=7/-4.861/137.252
![z7-timika-current](https://user-images.githubusercontent.com/42757252/53777143-df5b7500-3f2a-11e9-8e3a-8a57c2502233.png)

Without coast outline
![z7-timika-after](https://user-images.githubusercontent.com/42757252/53777152-e71b1980-3f2a-11e9-800a-e2cc32d4d489.png)

**Comment by matkoniecz** [comment]
I admit that in places with irregular coast (Timika, small islands) dropping border works well but in other places losing border makes map worse.

I am not sure what is preferable.

**Comment by imagico** [comment]
The question that arises is what the goal of #3065 is.  I consider it highly problematic in terms of mapper feedback (both due to the errors shown here and due to other issues it is very hard for mappers to understand how the appearance in the map relates to how things are mapped without an in depth knowledge how things are rendered).  And i also don't think it really serves as a usable example for other map styles because of inherent flaws in the method used and since parameters have been tuned to work around the various problems (like the 180 degree meridian artefacts, small islands, the wrong drawing order).

Independent of the specific technical issues having a special rendering style for coastlines at z0-z7 but not above indicates a specific low zoom design idea for the map.  But so far i don't see this.  Most map styles that draw coastlines or waterbody outlines in general with a line do so at large scales, not at small scales.

**Comment by jragusa** [comment]
Timika is probably not a good example : mappers extended coastline (https://www.openstreetmap.org/relation/8661308#map=13/-5.0509/137.4330) along rivers (https://www.openstreetmap.org/relation/8504221) which explains the weird pattern

**Comment by jeisenbe** [comment]
> Timika is probably not a good example : mappers extended coastline (https://www.openstreetmap.org/relation/8661308#map=13/-5.0509/137.4330) along rivers (https://www.openstreetmap.org/relation/8504221) which explains the weird pattern

Most of those are not rivers but tidal channels in mangroves. They are
filled with salt water and the flow changes direction with the tide.
Some of the channels are connected to rivers, but these sections of
the rivers are tidal estuaries, and should be mapped as outside of the
coastline.

But similar effects can be seen in areas with fjords and rias (flooded
glacial and river valleys), eg Norway and Cornwall, or in areas of
salt marsh with many tidal channels

https://wiki.openstreetmap.org/wiki/Proposed_Features/Coastline-River_transit_placement

**Comment by Adamant36** [comment]
> Timika is probably not a good example : mappers extended coastline (https://www.openstreetmap.org/relation/8661308#map=13/-5.0509/137.4330) along rivers (https://www.openstreetmap.org/relation/8504221) which explains the weird pattern

That happened to some places that I map where people extended the river area miles up streams. It was rather irritating. I think the blue outline would be good in cases where the coastline is mostly straight, but not in most of the examples here. Is there a way it can be toned down enough to not stand out as much or make it only apply to semi-straight coastlines (that would probably look inconsistent)?

**Comment by imagico** [comment]
> Most of those are not rivers but tidal channels in mangroves. They are filled with salt water and the flow changes direction with the tide.

Indeed.  Coastline placement in the OSM database is and should be a matter of local observable properties, not of what is considered cartographically desirable for a specific styling idea at a specific map scale.  This style should incentivize the former, not the latter.  A styling concept that emphasizes the significance of the coastline at low zooms but de-emphasizes it at high zooms does not really support this however.

The discussion here IMO should focus on the cartographic merits of #3065 w.r.t. to the goals we have.  It would be good to get the input from those who originally discussed this change (@matthijsmelissen, @kocio-pl and @pnorman).  There is not much in the PR discussion beyond the expression of subjective like for the change.

**Comment by pnorman** [comment]
I'm in favour of removal as this issue has brought up places where it fails badly.

**Comment by matthijsmelissen** [comment]
> The question that arises is what the goal of #3065 is. 

The goal is to increase contrast between water and land, which makes the map easier to read, especially around complex coastlines.

With respect to the proposed revert, I'm fairly neutral.

**Comment by matthijsmelissen** [comment]
A general point: personally I am against optimizing for 2x exports at the cost of the 1x version as long as openstreetmap.org renders the 1x version.

**Comment by imagico** [comment]
I don't think @jeisenbe showed the high resolution samples as a problem per se but to better illustrate problems observed at normal resolution.

If the aim is purely to increase contrast of the land-water boundary the line simplification does not appear to be useful.  A dark halo around land areas on the water side is something that can work nicely without the distortion inherent to geometry modification steps.

**Comment by kocio-pl** [comment]
For me current rendering improves clarity. Less details together with better visible line at lower zoom levels works great in most of the cases. Timika etc. is visually worse (but still acceptable for me) and conceptually it's highlighting the special data situation there (we claim there are some kind of small ocean islands, not part of the land).

**Comment by imagico** [comment]
@matthijsmelissen - what i still don't understand about #3065 is the reason why z0 to z7 and not above.  Was this because of interference with landcover rendering (which back then started at z8)?  That would mean #3458 was essentially incompatible to some extent with #3065.  The effect is relatively subtle though (some additional non-uniformity in the line where landcover is mapped near the coast) and it probably is even more so at z8/z9 so it seems doubtful this played a role.

**Comment by matthijsmelissen** [comment]
>  what i still don't understand about #3065 is the reason why z0 to z7 and not above.

It's because I wanted to avoid blue lines crossing rivers at the place where rivers run into the ocean. At z7 and lower this is not noticeable in most places because rivers at these zoom levels are very thin. From z8, some rivers start to become multiple pixels wide at the place they reach the see.

Landuse has nothing to do with this.

**Comment by imagico** [comment]
Ah, ok - thanks.

**Comment by jeisenbe** [comment]
It looks like @kocio-pl is still in favor of using generalization and outlines at low zoom, @matthijsmelissen is neutral, @imagico and @pnorman are opposed to the current rendering. 

I would suggest merging #3694 for now, which will remove the current simplification and outlines.

The option that would provide and outline with better results would be to use the generalized coastlines from http://openstreetmapdata.com/data/generalized-coastlines - which conveniently are provided for z0 to z8. These still show some problems in areas with many tidal channels, but they do not have any issues with missing islands, and the generalization process as a very nice cartographic result. However, they are even more simplified/generalized than the current low-zoom outlines.

But we need to decide on this soon, if we want these shapefiles to stay available. See: https://blog.jochentopf.com/2019-03-07-the-new-osmdata-service.html 

"We are currently not offering the Generalized coastlines, the Reduced waterbodies as raster mask and the Reduced waterbodies as polygons on the new service. We might bring them to the new service, but this depends on whether people actually need these. Please contact me if you currently use any of them."

Greece at z6 with generalized coastlines:
![z5-export2x-greece-generalized-coastlines-1zlate](https://user-images.githubusercontent.com/42757252/53300207-9e9e9480-3876-11e9-9e41-3e714cdc85e5.png)

Compare to z6 rendering with simplified water polygons (i.e. with PR #3694)
![z6-greece-ocean-simplified-water-polygons](https://user-images.githubusercontent.com/42757252/54169724-4347e580-44b7-11e9-9e1c-25937172ba97.png)

And here's z7 Wales with generalized coastlines (and landcover colors simplified as in PR #3670)
![z7-wales-4colors-generalized](https://user-images.githubusercontent.com/42757252/54169845-ca955900-44b7-11e9-8905-ac2d79a9cd95.png)

Compare to z7 Wales with current coastline style (same landcover colors as #3670):
![z7-wales-altcolors](https://user-images.githubusercontent.com/42757252/54169890-fe707e80-44b7-11e9-8a83-335b2db14707.png)

z3 Indonesia with generalized coastline and outline:
![z3-indonesia-generalized-outline](https://user-images.githubusercontent.com/42757252/54170043-76d73f80-44b8-11e9-99df-53c8add2ffc2.png)

z3 Indonesia current style:
![z3-indonesia-ocean-master](https://user-images.githubusercontent.com/42757252/54170039-7474e580-44b8-11e9-9178-e470cf4dd05c.png)

z3 Europe, generalized:
![z3-ocean-generalized](https://user-images.githubusercontent.com/42757252/54170184-0381fd80-44b9-11e9-8bcd-e1dc68b485e2.png)

z3 Europe, current style:
![z3-ocean-master](https://user-images.githubusercontent.com/42757252/54170201-11d01980-44b9-11e9-8a48-b6246b915dcc.png)

z3 Europe, after #3694 (without outline):
![z3-ocean-simplified-water-polygons](https://user-images.githubusercontent.com/42757252/54170293-7a1efb00-44b9-11e9-942f-5588c808eb9c.png)

**Comment by jeisenbe** [comment]
Here are some more comparisons with generalized coastline shapefiles vs current rendering and outline-less rendering.

Papua, Indonesia (as mentioned in the [comment above](https://github.com/gravitystorm/openstreetmap-carto/issues/3695#issuecomment-469513193))
https://www.openstreetmap.org/#map=7/-4.861/137.252
z6 current master
![z6-Papua-master](https://user-images.githubusercontent.com/42757252/54174846-ea367c80-44cb-11e9-8cf2-7badd2f71379.png)
z7 current master
![z7-timika-without-outline](https://user-images.githubusercontent.com/42757252/54174849-ee629a00-44cb-11e9-8b6f-239add26753e.png)

z6 without outline (after PR #3694)
![z6-papua-after](https://user-images.githubusercontent.com/42757252/54174857-f3274e00-44cb-11e9-853f-d37d0116b8c4.png)
z7 without outline (after PR #3694)
![z7-timika-without-outline](https://user-images.githubusercontent.com/42757252/54174861-f6bad500-44cb-11e9-8d0e-4b32e275c588.png)

z6 generalized coastline, thicker outline
![z5-export2x-timika-generalized-1xlate](https://user-images.githubusercontent.com/42757252/54174874-02a69700-44cc-11e9-8eb6-ae5dfb5ae265.png)
z7 generalized coastline, thicker outline
![z6-2x-export-timika-generalized-1late](https://user-images.githubusercontent.com/42757252/54174880-063a1e00-44cc-11e9-81c3-f7c56bd3dbfb.png)

z6 less generalized (using shapefiles 1 zoom level lower than designed)
![z6-timika-generalized-1zlate](https://user-images.githubusercontent.com/42757252/54174890-13efa380-44cc-11e9-9f34-4fe677fe1679.png)
z7 less generalized (using shapefiles 1 zoom level lower than designed)
![z7-timika-generalized-1zlate](https://user-images.githubusercontent.com/42757252/54174891-1651fd80-44cc-11e9-95cf-898a610d897d.png)

- This last option looks more similar to the current rendering with outline and simplification, by using the generalized shapefiles one zoom level sooner than intended, as an example of using less generalization.

**Comment by jeisenbe** [comment]
Scotland, UK:
z7 current master
![z7-scotland-master](https://user-images.githubusercontent.com/42757252/54174956-4bf6e680-44cc-11e9-810d-f35733300cc6.png)

z7 without outline (after PR #3694)
![z7-scotland-water-polygons](https://user-images.githubusercontent.com/42757252/54174987-616c1080-44cc-11e9-8c9c-1f1076ad50c3.png)

z7 generalized coastline, thicker outline
![z6-export2x-generalized-scotland-1zlate](https://user-images.githubusercontent.com/42757252/54175000-747ee080-44cc-11e9-850e-40b88a812aa0.png)

z7 less generalized (using shapefiles 1 zoom level lower than designed)
![z7-scotland-generalized-1zlate](https://user-images.githubusercontent.com/42757252/54175013-81033900-44cc-11e9-9231-1e91d9e70f13.png)

**Comment by Adamant36** [comment]
Its a trade off definitely. z6 without outline (after PR #3694) is harder to see without the outline. It's to bad water isn't a tag darker.

**Comment by imagico** [comment]
Note moving further processes from openstreetmapdata.com to the new platform is not a matter of urgency, it is a matter of available time to actually do so.  The generalized coastlines currently have the lowest priority in that regard.

One of the main advantages of the new platform is that different processes are more compartmentalized now so it is relatively easy for anyone to develop new processes within this framework and add them.  Anyone is welcome to do so on https://github.com/fossgis/osmdata.

The other question is of course how much pre-processing of data is actually desirable.  I am generally in favor of this where it is needed for constructive mapper feedback and a process exists to do so but i don't really think this style is a place to add complexity in pre-processing _just for eye candy_ so to speak.

**Comment by jeisenbe** [comment]
**Shikoku, Japan: current master**
z5 before
![z5-shikoku-master](https://user-images.githubusercontent.com/42757252/54169288-71c4c100-44b5-11e9-82b0-56e7f8ecb93c.png)

z6 before
![z6-shikoku-master](https://user-images.githubusercontent.com/42757252/54169291-72f5ee00-44b5-11e9-841f-c3923225d375.png)

z7 before
![z7-shikoku-master](https://user-images.githubusercontent.com/42757252/54169302-80ab7380-44b5-11e9-93c1-bfc34e8490e3.png)

**After PR 3694 (no outline)**
z5 no outline
![z5-shikoku-ocean-polygons](https://user-images.githubusercontent.com/42757252/54253790-2e8d4f80-4594-11e9-9760-d0bb97dbe04c.png)

z6 no outline
![z6-shikoku-ocean-polygons](https://user-images.githubusercontent.com/42757252/54253809-45cc3d00-4594-11e9-93fc-09a71d925699.png)

z7 no outline
![z7-shikoku-ocean-polygons](https://user-images.githubusercontent.com/42757252/54254559-88434900-4597-11e9-9847-49eeda66462e.png)

**Generalized coastlines** - [1 zoom level late]
z5 generalized
![z5-shikoku-generalized](https://user-images.githubusercontent.com/42757252/54253789-2e8d4f80-4594-11e9-9852-156b0530d242.png)

z6 generalized
![z6-shikoku-generalized](https://user-images.githubusercontent.com/42757252/54253783-2af9c880-4594-11e9-8636-ba63076f7084.png)

z7 generalized
![z7-shikoku-generalized](https://user-images.githubusercontent.com/42757252/54253794-3220d680-4594-11e9-8979-db5a02c1923b.png)

**Comment by jeisenbe** [comment]
Estonia has an interesting coastline and very complete landcover mapping:

### Current rendering:
z5
![z5-estonia-master](https://user-images.githubusercontent.com/42757252/54327658-3b6e7980-464e-11e9-9709-38140595dac6.png)
z6
![z6-estonia-master](https://user-images.githubusercontent.com/42757252/54327670-3f020080-464e-11e9-9e76-f74bb8968fa2.png)
z7
![z7-estonia-master](https://user-images.githubusercontent.com/42757252/54327675-432e1e00-464e-11e9-9ab4-9debdbbd3475.png)

### Water polygons, no outline:
z5
![z5-estonia-water-polygons](https://user-images.githubusercontent.com/42757252/54327666-3e696a00-464e-11e9-9075-b5981b33ad31.png)
z6
![z6-estonia-water-polygons](https://user-images.githubusercontent.com/42757252/54327674-432e1e00-464e-11e9-9a9e-4289396320d4.png)
z7
![z7-estonia-water-polygons](https://user-images.githubusercontent.com/42757252/54327672-40cbc400-464e-11e9-8c9a-a78be65334bc.png)

### Generalized coastline:
z5
![z5-estonia-generalized-original](https://user-images.githubusercontent.com/42757252/54326589-2ee82200-464a-11e9-8499-09e8ca6781ec.png)
z6
![z6-estonia-generalized-original](https://user-images.githubusercontent.com/42757252/54326588-2db6f500-464a-11e9-8b50-21316cfba4ad.png)
z7
![z7-estonia-generalized-original](https://user-images.githubusercontent.com/42757252/54326582-298ad780-464a-11e9-9f4f-264304bbfc69.png)

### Less generalized (by using generalized shapefiles 1 zoom level earlier)
z5
![z5-estonia-generalized-less](https://user-images.githubusercontent.com/42757252/54328345-e54f0580-4650-11e9-8015-b95299e6121e.png)
z6
![z6-estonia-generalized-less](https://user-images.githubusercontent.com/42757252/54328347-e5e79c00-4650-11e9-8079-64e864f47b96.png)
z7
![z7-estonia-generalized-less](https://user-images.githubusercontent.com/42757252/54328349-e7b15f80-4650-11e9-94f4-352aadc705ad.png)

**Comment by pnorman** [comment]
If we want to discuss generalized coastline that should be its own issue and we can then keep this confined to just the switch to ocean polygons.

**Comment by jeisenbe** [comment]
Now that #3694 is merged, we should decide if the outline is important enough to be worth the extra complexity and the dependence on external shapefiles. One objection to the current outlines is that they did not extend to lower zoom levels. This was because the line would look strange when cutting across rivers and estuaries.

However, this issue has a solution: if we change rivers to a darker color as planned, then the outline for marine waters can be the same color as the new river color. This way, the marine water outline blends perfectly with the river water color. 

Here are examples with the 3-color water system, using the generalized coastlines and shapefiles for < z8 and the simplified and split water polygons and coastline shapefiles for z9 and up:

**Anchorage, Alaska z7 current**
![z7-anchorage-before](https://user-images.githubusercontent.com/42757252/54918706-51c7df80-4f42-11e9-8e31-d945746af277.png)
**After:** With 3 colors for water (lighter ocean, darker rivers) and marine water outline matching river color:
![z7-anchorage-after](https://user-images.githubusercontent.com/42757252/54918745-702ddb00-4f42-11e9-8b62-b141512e44e4.png)

z8 current
![z8-anchorage-before](https://user-images.githubusercontent.com/42757252/54918766-78861600-4f42-11e9-889c-90537bdd786f.png)
z8 after
![z8-anchorage-after](https://user-images.githubusercontent.com/42757252/54918779-7e7bf700-4f42-11e9-83bd-f92015638744.png)

z9 current
![z9-anchorage-before](https://user-images.githubusercontent.com/42757252/54918787-83d94180-4f42-11e9-9a2f-9a867e673508.png)
z9 after
![z9-anchorage-after](https://user-images.githubusercontent.com/42757252/54918797-876cc880-4f42-11e9-835d-03dba385b2ab.png)

**Bremen, Germany**
- This is an example where the coastline extends far up the river till the tidal section ends at a weir or dam, in the city of Bremen
z11 before
![z11-bremen-before](https://user-images.githubusercontent.com/42757252/54918899-bd11b180-4f42-11e9-86f8-1e90fa8297d5.png)
z11 after
![z11-bremen-after](https://user-images.githubusercontent.com/42757252/54918907-c1d66580-4f42-11e9-9b18-7b6b68a629e8.png)

z12 before
![z12-bremen-before](https://user-images.githubusercontent.com/42757252/54918914-c69b1980-4f42-11e9-801e-99d84c27d4b0.png)
z12 after
![z12-bremen-after](https://user-images.githubusercontent.com/42757252/54918920-cac73700-4f42-11e9-839a-a431c1453fea.png)

z14 before
![z14-weser-before](https://user-images.githubusercontent.com/42757252/54918879-b4b97680-4f42-11e9-966e-84eb942ff6f2.png)
z14 after
![z14-weser-bremen-after](https://user-images.githubusercontent.com/42757252/54918890-b7b46700-4f42-11e9-9548-28d9fb3f2f2c.png)

**Comment by imagico** [comment]
That does not seem a good idea because it would introduce a water bias and would give confusing feedback to mappers showing the ocean outline in a way that looks like a river geometry.

I would suggest to close this issue because the specific problems discussed here no more apply after #3694 and instead would open specific issues on problems we have and address them in the best way possible.  But it would probably make sense to put priority on those things that have been waiting for the move to ocean polygons and layer-reordering.

One thing i hope this subject has shown to everyone is that you have to be careful making design choices that severely limit your options w.r.t. other changes and that maneuver the style into a position where you can't improve anything any more without sacrificing central goals or reverting previous choices made.

What i would like to see - and which kind of goes into a similar direction as your samples is an edge emphasized (in terms of color) rendering of oceans that does not introduce a land or water bias and that does work independently of geometry simplification.  This is not trivial, it probably requires some comp-op magic similar to what @sommerluk developed for the unpaved roads.  And i am not sure if the results would actually be desirable to have in this style.  But it would definitely be something to try out.

**Comment by jeisenbe** [comment]
**Zeeland, the Netherlands**
- Currently it's not easy to see where the coastline is placed in this area, below z7
z7 zoomed in, current
![z7-zeeland-current-2x](https://user-images.githubusercontent.com/42757252/54921618-8c814600-4f49-11e9-8602-2fe655f7360a.png)
z7 after (with landcover and water color changes in addition)
![z7-zeeland-after-2x](https://user-images.githubusercontent.com/42757252/54921625-91de9080-4f49-11e9-9baa-b8b639ab9d0b.png)

z9 Zeeland current
![z9-zeeland-before](https://user-images.githubusercontent.com/42757252/54921633-973bdb00-4f49-11e9-9a55-17f672e08e06.png)

z9 after
![z9-zeeland-after](https://user-images.githubusercontent.com/42757252/54921818-0fa29c00-4f4a-11e9-93f2-4d30d3cd285f.png)

z10 current
![Uploading z10-zeeland-before.png…]()

z10 after
![Uploading z10-zeeland-after.png…]()

The position and width of the outline could be tweaked, but I'm just showing these images to give a general idea of what it might look like.

**Comment by jeisenbe** [comment]
>  it would introduce a water bias
Do you mean that it would cause narrow sections of coastline to appear wider than they are in reality, eg tidal channels, narrow bays? We do show narrow rivers and streams as wider than their actual sizes at these zoom levels.

> ... confusing feedback to mappers showing the ocean outline in a way that looks like a river geometry.

Yes, that is an issue

> an edge emphasized (in terms of color) rendering of oceans that does not introduce a land or water bias

That would be great. It seems like having a gradient that gets lighter with increasing distance from the edge of the polygon would be a great way to render the marine environment; this is frequently seen on historic paper maps. 

 I couldn't find any documentation of this in Mapnik.

**Comment by imagico** [comment]
> Do you mean that it would cause narrow sections of coastline to appear wider than they are in reality, eg tidal channels, narrow bays? We do show narrow rivers and streams as wider than their actual sizes at these zoom levels.

I mean the outline drawing indicating the extent of the water area being larger than as it is mapped systematically decreasing the size of land and increasing the size of water in rendering.  This is different from centerline drawing of waterways where the line signature is never meant to indicate the water covered area.

> I couldn't find any documentation of this in Mapnik.

Mapnik has no native support for gradients, you'd have to approximate such with multiple lines in different colors.

**Comment by pnorman** [comment]
> Mapnik has no native support for gradients, you'd have to approximate such with multiple lines in different colors.

I'd do it with MSS similar to how we do nature reserves, but no transparency.

**Comment by imagico** [comment]
Yes, but nature reserves use line-offset and that would depend on unsplit polygons which would lead to the same constraints as we had with the outlines and line simplification.

**Comment by jeisenbe** [comment]
C

**Comment by jeisenbe** [comment]
> Yes, but nature reserves use line-offset and that would depend on unsplit polygons which would lead to the same constraints as we had with the outlines and line simplification.

We can use the coastline linestrings for the high zoom levels, but it would be idea to have slightly generalized coastlines up to z9. Right now the generalized coastlines available on openstreetmapdata.com end at z8. It would be perfect to have a transitional level at z8. Then we would not need to use `simplified-water-polygons-split`. 

Another option, although rather wasteful, is to use the simplified-land-polygons-complete for the coastline outline, and the water polygons to render the water fill, at z9; that's what I did in the examples above. But some `comp-op` sorcery would be necessary if it's essential that the outline not extend outside of the actual area of water, whichever shapefiles are used.

**Comment by jeisenbe** [comment]
I've tried several different pseudo-gradients by using line-offset and different line widths for each color, eg:

```
.coastline {
  [zoom >= 1] {
        c/line-width: 2.0;
        c/line-color: @water-color;
        c/line-offset: 1.0;
        b/line-width: 1.0;
        b/line-color: @river-color;
        b/line-offset: 0.5;
```

I haven't yet found the right `comp-op` option to get the rendering perfect, but I can show what it would like, ideally. It looks like we would need to double-up on shapefiles (using separate shapefiles for the water polygons and the coastline). Should I continue to work on this? 

**2 pixel wide gradient:**
z5 Alaska 2px
![z5-yukon-river-2pixel-gradient](https://user-images.githubusercontent.com/42757252/55559778-e512c900-5729-11e9-80df-a516880fe3f6.png)

z5 Great Lakes 2px
![z5-great-lakes-2pixel](https://user-images.githubusercontent.com/42757252/55559806-f78d0280-5729-11e9-88aa-450a09175749.png)

z6 Amazon 2px
![z6-amazon-river-2pixel-gradient](https://user-images.githubusercontent.com/42757252/55559834-01af0100-572a-11e9-9fb0-f5e3c807223e.png)

z6 Greece 2px
![z6-greece-2pixel-gradient](https://user-images.githubusercontent.com/42757252/55560341-fc05eb00-572a-11e9-86fa-41ac6d3c0c45.png)

z13 Bremen 2px
![z13-bremen-2pixel-gradient](https://user-images.githubusercontent.com/42757252/55559340-f7d8ce00-5728-11e9-9ece-f9c50f4d8df2.png)
z14
![z14-bremen-weser-2pixel-gradient](https://user-images.githubusercontent.com/42757252/55559344-f9a29180-5728-11e9-88cf-70862e3b0940.png)

**1 pixel wide gradient** (Half-pixel steps)
z3 Europe 1px
![z3-europe-1pixel-gradient](https://user-images.githubusercontent.com/42757252/55559689-b72d8480-5729-11e9-9bee-33c9d252d92f.png)

z5 Alaska 1px
![z5-yukon-river-1pixel-gradient](https://user-images.githubusercontent.com/42757252/55559896-1d1a0c00-572a-11e9-8b1b-4bf3431abf26.png)

z5 Great Lakes 1px
![z5-great-lakes-1px-gradient](https://user-images.githubusercontent.com/42757252/55559915-21dec000-572a-11e9-8568-74a5469b82ac.png)

z6 Amazon 1px
![z6-amazon-1pixel-gradient](https://user-images.githubusercontent.com/42757252/55560035-58b4d600-572a-11e9-93e2-4065ba9c9669.png)

z6 Greece 1px
![z6-greece-1pixel-gradient](https://user-images.githubusercontent.com/42757252/55560349-00320880-572b-11e9-8ece-2e135072b3f1.png)

z13 Bremen 1px
![z13-bremen-1pixel-gradient](https://user-images.githubusercontent.com/42757252/55559334-f5767400-5728-11e9-97d4-0b9e01c521bf.png)
z14
![z14-bremen-1pixel-gradient](https://user-images.githubusercontent.com/42757252/55559330-f14a5680-5728-11e9-89ac-13ee01c8c22b.png)


(I also tried a **4 pixel wide gradient**, but I don't think it looks good; small islands are over-emphasized)
z3 Europe 4px
![z3-europe-ocean-gradient](https://user-images.githubusercontent.com/42757252/55559575-77669d00-5729-11e9-8f6d-1eba608e59dd.png)

z5 Alaska 4px
![z5-yukon-river-gradient](https://user-images.githubusercontent.com/42757252/55560058-636f6b00-572a-11e9-94fc-ca3ccf6d19c2.png)

**Comment by imagico** [comment]
Yes, that goes in the direction i had in mind.  But as said you can't use line-offset because the lines are split and this will result in artefacts at the splits: Either gaps or overlaps between the offset segments depending on if the coastline is concave of convex at this point.  And you also get further distortion where the offset line drawing of one coastline geometry is drawn over another land feature.

Anyway - the logical next step to proceed with the coastlines is to do the layer re-ordering.  Anything else like different coloring of different water features or this would come afterwards.

**Comment by pnorman** [comment]
> Anyway - the logical next step to proceed with the coastlines is to do the layer re-ordering. Anything else like different coloring of different water features or this would come afterwards.

+1. There were other ideas we had for improvements after switching to water polygons.

**Comment by jeisenbe** [comment]
These are the gradients formed using `dst-out` and `dst-over`. I made them back on 5-April-2019 but had decided not to work on it further at the time. These tests also used different colors for rivers, lakes and ocean. As you see, it works:

![z12-timika-rivers-dst-over](https://user-images.githubusercontent.com/42757252/63924521-637a5980-ca83-11e9-9165-f1d234a4fcb4.png)

![z6-amazon-2px-gradient-dst-over](https://user-images.githubusercontent.com/42757252/63924714-b6eca780-ca83-11e9-9875-19bd2bc432fa.png)

![z5-great-lakes-2px-dst-over](https://user-images.githubusercontent.com/42757252/63924716-b81dd480-ca83-11e9-9deb-e40adc861641.png)

**Comment by jeisenbe** [comment]
This issue was solved by #3694 so I will open a new issue discuss adding back an outline to the ocean polygons, which requires relayering and using comp-op as discussed in #3854
